File: bindings/java/src/main/java/com/strling/core/IR.java
package com.strling.core;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * STRling Intermediate Representation (IR) Node Definitions.
 * 
 * <p>This module defines the complete set of IR node classes that represent
 * language-agnostic regex constructs. The IR serves as an intermediate layer
 * between the parsed AST and the target-specific emitters (e.g., PCRE2).</p>
 * 
 * <p>IR nodes are designed to be:</p>
 * <ul>
 *   <li>Simple and composable</li>
 *   <li>Easy to serialize (via toDict methods)</li>
 *   <li>Independent of any specific regex flavor</li>
 *   <li>Optimized for transformation and analysis</li>
 * </ul>
 * 
 * <p>Each IR node corresponds to a fundamental regex operation (alternation,
 * sequencing, character classes, quantification, etc.) and can be serialized
 * to a dictionary representation for further processing or debugging.</p>
 */
public class IR {
    
    /**
     * Base class for all IR operations.
     * 
     * <p>All IR nodes extend this base class and must implement the toDict() method
     * for serialization to a dictionary/object representation.</p>
     */
    public static abstract class IROp {
        /**
         * Serializes the IR node to a dictionary representation.
         * 
         * @return The dictionary representation of this IR node
         */
        public abstract Map<String, Object> toDict();
    }
    
    /**
     * Represents an alternation (OR) operation in the IR.
     * 
     * <p>Matches any one of the provided branches. Equivalent to the | operator
     * in traditional regex syntax.</p>
     */
    public static class IRAlt extends IROp {
        public List<IROp> branches;
        
        /**
         * Creates an alternation IR node.
         *
         * @param branches Array of IR nodes representing the alternative branches
         */
        public IRAlt(List<IROp> branches) {
            this.branches = branches;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Alt");
            List<Map<String, Object>> branchDicts = new ArrayList<>();
            for (IROp branch : branches) {
                branchDicts.add(branch.toDict());
            }
            map.put("branches", branchDicts);
            return map;
        }
    }
    
    /**
     * Represents a sequence (concatenation) operation in the IR.
     * 
     * <p>Matches all parts in order. Equivalent to concatenating patterns
     * in traditional regex syntax.</p>
     */
    public static class IRSeq extends IROp {
        public List<IROp> parts;
        
        /**
         * Creates a sequence IR node.
         *
         * @param parts Array of IR nodes to be matched sequentially
         */
        public IRSeq(List<IROp> parts) {
            this.parts = parts;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Seq");
            List<Map<String, Object>> partDicts = new ArrayList<>();
            for (IROp part : parts) {
                partDicts.add(part.toDict());
            }
            map.put("parts", partDicts);
            return map;
        }
    }
    
    /**
     * Represents a literal string in the IR.
     *
     * <p>Matches the exact string value with all characters treated as literals
     * (no special regex meaning).</p>
     */
    public static class IRLit extends IROp {
        public String value;
        
        /**
         * Creates a literal IR node.
         *
         * @param value The literal string to match
         */
        public IRLit(String value) {
            this.value = value;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Lit");
            map.put("value", value);
            return map;
        }
    }
    
    /**
     * Represents the "any character" wildcard in the IR.
     *
     * <p>Matches any single character (behavior may depend on flags like dotall).
     * Equivalent to the . metacharacter in traditional regex.</p>
     */
    public static class IRDot extends IROp {
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Dot");
            return map;
        }
    }
    
    /**
     * Represents an anchor (position assertion) in the IR.
     *
     * <p>Matches a position rather than a character. Common anchors include
     * start of line (^), end of line ($), word boundaries (\b), etc.</p>
     */
    public static class IRAnchor extends IROp {
        public String at;
        
        /**
         * Creates an anchor IR node.
         *
         * @param at The anchor type (e.g., "Start", "End", "WordBoundary")
         */
        public IRAnchor(String at) {
            this.at = at;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Anchor");
            map.put("at", at);
            return map;
        }
    }
    
    /**
     * Base class for character class items.
     *
     * <p>Character class items represent the individual components that can appear
     * inside a character class (ranges, literals, or escape sequences).</p>
     */
    public static abstract class IRClassItem {
        /**
         * Serializes the character class item to a dictionary representation.
         * 
         * @return Map containing the item's serialized representation
         */
        public abstract Map<String, Object> toDict();
    }
    
    /**
     * Represents a character range in a character class.
     *
     * <p>Matches any character between fromCh and toCh inclusive (e.g., a-z, 0-9).</p>
     */
    public static class IRClassRange extends IRClassItem {
        public String fromCh;
        public String toCh;
        
        /**
         * Creates a character range item.
         *
         * @param fromCh The starting character of the range
         * @param toCh The ending character of the range
         */
        public IRClassRange(String fromCh, String toCh) {
            this.fromCh = fromCh;
            this.toCh = toCh;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Range");
            map.put("from", fromCh);
            map.put("to", toCh);
            return map;
        }
    }
    
    /**
     * Represents a single literal character in a character class.
     *
     * <p>Matches exactly the specified character.</p>
     */
    public static class IRClassLiteral extends IRClassItem {
        public String ch;
        
        /**
         * Creates a literal character item.
         *
         * @param ch The literal character to match
         */
        public IRClassLiteral(String ch) {
            this.ch = ch;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Char");
            map.put("char", ch);
            return map;
        }
    }
    
    /**
     * Represents an escape sequence in a character class.
     *
     * <p>Handles special character class escapes like \d, \w, \s, or Unicode
     * property escapes like \p{Letter}.</p>
     */
    public static class IRClassEscape extends IRClassItem {
        public String type;
        public String property;
        
        /**
         * Creates an escape sequence item.
         *
         * @param type The escape type (e.g., "Digit", "Word", "Whitespace", "UnicodeProperty")
         */
        public IRClassEscape(String type) {
            this(type, null);
        }
        
        /**
         * Creates an escape sequence item.
         *
         * @param type The escape type (e.g., "Digit", "Word", "Whitespace", "UnicodeProperty")
         * @param property The Unicode property name if type is "UnicodeProperty"
         */
        public IRClassEscape(String type, String property) {
            this.type = type;
            this.property = property;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Esc");
            map.put("type", type);
            if (property != null) {
                map.put("property", property);
            }
            return map;
        }
    }
    
    /**
     * Represents a character class in the IR.
     *
     * <p>Matches any single character from the set of allowed characters or ranges.
     * Can be negated to match any character NOT in the set.</p>
     */
    public static class IRCharClass extends IROp {
        public boolean negated;
        public List<IRClassItem> items;
        
        /**
         * Creates a character class IR node.
         *
         * @param negated Whether the character class is negated (e.g., [^...])
         * @param items List of character class items (ranges, literals, escapes)
         */
        public IRCharClass(boolean negated, List<IRClassItem> items) {
            this.negated = negated;
            this.items = items;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "CharClass");
            map.put("negated", negated);
            List<Map<String, Object>> itemDicts = new ArrayList<>();
            for (IRClassItem item : items) {
                itemDicts.add(item.toDict());
            }
            map.put("items", itemDicts);
            return map;
        }
    }
    
    /**
     * Represents a quantifier (repetition) in the IR.
     *
     * <p>Specifies how many times the child pattern should be matched.
     * Supports greedy, lazy, and possessive matching modes.</p>
     */
    public static class IRQuant extends IROp {
        public IROp child;
        public int min;
        /**
         * Maximum repetitions (number), or 0 for unbounded (infinity).
         */
        public Object max; // int or 0 for "Inf"
        /**
         * Quantifier mode: "Greedy", "Lazy", or "Possessive".
         */
        public String mode;
        
        /**
         * Creates a quantifier IR node.
         *
         * @param child The IR node to be quantified
         * @param min Minimum number of repetitions
         * @param max Maximum number of repetitions (number or 0 for unlimited)
         * @param mode Matching mode: "Greedy", "Lazy", or "Possessive"
         */
        public IRQuant(IROp child, int min, Object max, String mode) {
            this.child = child;
            this.min = min;
            this.max = max;
            this.mode = mode;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Quant");
            map.put("child", child.toDict());
            map.put("min", min);
            map.put("max", max);
            map.put("mode", mode);
            return map;
        }
    }
    
    /**
     * Represents a group (capturing or non-capturing) in the IR.
     *
     * <p>Groups are used for capturing matched text, applying quantifiers to
     * multiple elements, or creating atomic groups for backtracking control.</p>
     */
    public static class IRGroup extends IROp {
        public boolean capturing;
        public IROp body;
        public String name;
        public Boolean atomic;
        
        /**
         * Creates a group IR node.
         *
         * @param capturing Whether this group captures its match
         * @param body The IR node contained within the group
         */
        public IRGroup(boolean capturing, IROp body) {
            this(capturing, body, null, null);
        }
        
        /**
         * Creates a group IR node.
         *
         * @param capturing Whether this group captures its match
         * @param body The IR node contained within the group
         * @param name Optional name for named capture groups
         */
        public IRGroup(boolean capturing, IROp body, String name) {
            this(capturing, body, name, null);
        }
        
        /**
         * Creates a group IR node.
         *
         * @param capturing Whether this group captures its match
         * @param body The IR node contained within the group
         * @param name Optional name for named capture groups
         * @param atomic Whether this is an atomic group (prevents backtracking)
         */
        public IRGroup(boolean capturing, IROp body, String name, Boolean atomic) {
            this.capturing = capturing;
            this.body = body;
            this.name = name;
            this.atomic = atomic;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Group");
            map.put("capturing", capturing);
            map.put("body", body.toDict());
            if (name != null) {
                map.put("name", name);
            }
            if (atomic != null && atomic) {
                map.put("atomic", atomic);
            }
            return map;
        }
    }
    
    /**
     * Represents a backreference in the IR.
     *
     * <p>References a previously captured group by index or name, matching
     * the exact same text that was captured by that group.</p>
     */
    public static class IRBackref extends IROp {
        public Integer byIndex;
        public String byName;
        
        /**
         * Creates a backreference IR node.
         */
        public IRBackref() {
            this(null, null);
        }
        
        /**
         * Creates a backreference IR node.
         *
         * @param byIndex The numeric group index to reference
         * @param byName The name of the group to reference
         */
        public IRBackref(Integer byIndex, String byName) {
            this.byIndex = byIndex;
            this.byName = byName;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Backref");
            if (byIndex != null) {
                map.put("byIndex", byIndex);
            }
            if (byName != null) {
                map.put("byName", byName);
            }
            return map;
        }
    }
    
    /**
     * Represents a lookaround assertion in the IR.
     *
     * <p>Lookarounds are zero-width assertions that check for pattern presence
     * or absence without consuming characters. They can look ahead or behind,
     * and can be positive or negative.</p>
     */
    public static class IRLook extends IROp {
        public String dir;
        public boolean neg;
        public IROp body;
        
        /**
         * Creates a lookaround IR node.
         *
         * @param dir Direction: "Ahead" for lookahead or "Behind" for lookbehind
         * @param neg Whether this is a negative assertion (pattern must NOT match)
         * @param body The IR node to look for
         */
        public IRLook(String dir, boolean neg, IROp body) {
            this.dir = dir;
            this.neg = neg;
            this.body = body;
        }
        
        @Override
        public Map<String, Object> toDict() {
            Map<String, Object> map = new HashMap<>();
            map.put("ir", "Look");
            map.put("dir", dir);
            map.put("neg", neg);
            map.put("body", body.toDict());
            return map;
        }
    }
}

File: bindings/swift/Sources/STRling/Core/IR.swift
/// STRling Intermediate Representation (IR) Node Definitions
///
/// This module defines the complete set of IR node classes that represent
/// language-agnostic regex constructs. The IR serves as an intermediate layer
/// between the parsed AST and the target-specific emitters (e.g., PCRE2).
///
/// IR nodes are designed to be:
///   - Simple and composable
///   - Easy to serialize (via toDict methods)
///   - Independent of any specific regex flavor
///   - Optimized for transformation and analysis
///
/// Each IR node corresponds to a fundamental regex operation (alternation,
/// sequencing, character classes, quantification, etc.) and can be serialized
/// to a dictionary representation for further processing or debugging.

import Foundation

// MARK: - IR Class Items

/// Base protocol for IR character class items
public protocol IRClassItem {
    /// Convert the class item to dictionary representation
    func toDict() -> [String: Any]
}

/// Represents a character range in IR character class (e.g., a-z)
public struct IRClassRange: IRClassItem {
    /// The starting character of the range
    public let fromCh: String
    
    /// The ending character of the range
    public let toCh: String
    
    /// Initialize an IR character range
    ///
    /// - Parameters:
    ///   - fromCh: The starting character
    ///   - toCh: The ending character
    public init(fromCh: String, toCh: String) {
        self.fromCh = fromCh
        self.toCh = toCh
    }
    
    public func toDict() -> [String: Any] {
        return [
            "ir": "Range",
            "from": fromCh,
            "to": toCh
        ]
    }
}

/// Represents a literal character in IR character class
public struct IRClassLiteral: IRClassItem {
    /// The literal character
    public let ch: String
    
    /// Initialize an IR character literal
    ///
    /// - Parameter ch: The literal character
    public init(ch: String) {
        self.ch = ch
    }
    
    public func toDict() -> [String: Any] {
        return [
            "ir": "Char",
            "char": ch
        ]
    }
}

/// Represents an escape sequence in IR character class (e.g., \d, \w, \s, \p{...})
public struct IRClassEscape: IRClassItem {
    /// The type of escape (d, D, w, W, s, S, p, P)
    public let type: String
    
    /// The Unicode property name (for \p and \P escapes)
    public let property: String?
    
    /// Initialize an IR class escape
    ///
    /// - Parameters:
    ///   - type: The type of escape
    ///   - property: Optional Unicode property name
    public init(type: String, property: String? = nil) {
        self.type = type
        self.property = property
    }
    
    public func toDict() -> [String: Any] {
        var d: [String: Any] = [
            "ir": "Esc",
            "type": type
        ]
        if let property = property {
            d["property"] = property
        }
        return d
    }
}

// MARK: - IR Operations

/// Enum representing all possible IR operation types.
///
/// This enum uses associated values to provide type-safe pattern matching
/// and easy traversal of the IR, following Swift best practices.
public indirect enum IROp {
    /// Alternation operation (OR)
    case alt(IRAlt)
    
    /// Sequence operation (concatenation)
    case seq(IRSeq)
    
    /// Literal string operation
    case lit(IRLit)
    
    /// Dot operation (matches any character)
    case dot(IRDot)
    
    /// Anchor operation (position assertion)
    case anchor(IRAnchor)
    
    /// Character class operation
    case charClass(IRCharClass)
    
    /// Quantifier operation
    case quant(IRQuant)
    
    /// Group operation
    case group(IRGroup)
    
    /// Backreference operation
    case backref(IRBackref)
    
    /// Lookaround operation
    case look(IRLook)
    
    /// Serialize the IR node to a dictionary representation.
    ///
    /// - Returns: The dictionary representation of this IR node
    public func toDict() -> [String: Any] {
        switch self {
        case .alt(let node):
            return node.toDict()
        case .seq(let node):
            return node.toDict()
        case .lit(let node):
            return node.toDict()
        case .dot(let node):
            return node.toDict()
        case .anchor(let node):
            return node.toDict()
        case .charClass(let node):
            return node.toDict()
        case .quant(let node):
            return node.toDict()
        case .group(let node):
            return node.toDict()
        case .backref(let node):
            return node.toDict()
        case .look(let node):
            return node.toDict()
        }
    }
}

/// Represents an alternation (OR) operation in the IR.
///
/// Matches any one of the provided branches. Equivalent to the | operator
/// in traditional regex syntax.
public struct IRAlt {
    /// The alternative branches
    public let branches: [IROp]
    
    /// Initialize an IR alternation
    ///
    /// - Parameter branches: The alternative branches
    public init(branches: [IROp]) {
        self.branches = branches
    }
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        return [
            "ir": "Alt",
            "branches": branches.map { $0.toDict() }
        ]
    }
}

/// Represents a sequence (concatenation) operation in the IR.
public struct IRSeq {
    /// The parts of the sequence
    public let parts: [IROp]
    
    /// Initialize an IR sequence
    ///
    /// - Parameter parts: The parts to concatenate
    public init(parts: [IROp]) {
        self.parts = parts
    }
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        return [
            "ir": "Seq",
            "parts": parts.map { $0.toDict() }
        ]
    }
}

/// Represents a literal string operation in the IR.
public struct IRLit {
    /// The literal string value
    public let value: String
    
    /// Initialize an IR literal
    ///
    /// - Parameter value: The literal string
    public init(value: String) {
        self.value = value
    }
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        return [
            "ir": "Lit",
            "value": value
        ]
    }
}

/// Represents a dot (.) operation in the IR - matches any character.
public struct IRDot {
    /// Initialize an IR dot
    public init() {}
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        return ["ir": "Dot"]
    }
}

/// Represents an anchor (position assertion) operation in the IR.
public struct IRAnchor {
    /// The anchor type
    public let at: String
    
    /// Initialize an IR anchor
    ///
    /// - Parameter at: The anchor type
    public init(at: String) {
        self.at = at
    }
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        return [
            "ir": "Anchor",
            "at": at
        ]
    }
}

/// Represents a character class operation in the IR.
public struct IRCharClass {
    /// Whether the character class is negated
    public let negated: Bool
    
    /// The items in the character class
    public let items: [IRClassItem]
    
    /// Initialize an IR character class
    ///
    /// - Parameters:
    ///   - negated: Whether the class is negated
    ///   - items: The items in the class
    public init(negated: Bool, items: [IRClassItem]) {
        self.negated = negated
        self.items = items
    }
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        return [
            "ir": "CharClass",
            "negated": negated,
            "items": items.map { $0.toDict() }
        ]
    }
}

/// Represents a quantifier operation in the IR.
public struct IRQuant {
    /// The child operation being quantified
    public let child: IROp
    
    /// Minimum number of repetitions
    public let min: Int
    
    /// Maximum number of repetitions
    public let max: IRQuantMax
    
    /// Quantifier mode: "Greedy", "Lazy", or "Possessive"
    public let mode: String
    
    /// Initialize an IR quantifier
    ///
    /// - Parameters:
    ///   - child: The operation to quantify
    ///   - min: Minimum repetitions
    ///   - max: Maximum repetitions
    ///   - mode: The quantifier mode
    public init(child: IROp, min: Int, max: IRQuantMax, mode: String) {
        self.child = child
        self.min = min
        self.max = max
        self.mode = mode
    }
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        return [
            "ir": "Quant",
            "child": child.toDict(),
            "min": min,
            "max": max.toValue(),
            "mode": mode
        ]
    }
}

/// Represents the maximum value for an IR quantifier
public enum IRQuantMax {
    /// Specific number of repetitions
    case count(Int)
    
    /// Infinite repetitions
    case inf
    
    /// Convert to value (Int or String "Inf")
    func toValue() -> Any {
        switch self {
        case .count(let n):
            return n
        case .inf:
            return "Inf"
        }
    }
}

/// Represents a group operation in the IR.
public struct IRGroup {
    /// Whether the group is capturing
    public let capturing: Bool
    
    /// The body of the group
    public let body: IROp
    
    /// Optional group name (for named captures)
    public let name: String?
    
    /// Whether the group is atomic
    public let atomic: Bool?
    
    /// Initialize an IR group
    ///
    /// - Parameters:
    ///   - capturing: Whether the group captures
    ///   - body: The group body
    ///   - name: Optional group name
    ///   - atomic: Optional atomic flag
    public init(capturing: Bool, body: IROp, name: String? = nil, atomic: Bool? = nil) {
        self.capturing = capturing
        self.body = body
        self.name = name
        self.atomic = atomic
    }
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        var d: [String: Any] = [
            "ir": "Group",
            "capturing": capturing,
            "body": body.toDict()
        ]
        if let name = name {
            d["name"] = name
        }
        if let atomic = atomic {
            d["atomic"] = atomic
        }
        return d
    }
}

/// Represents a backreference operation in the IR.
public struct IRBackref {
    /// Reference by numeric index
    public let byIndex: Int?
    
    /// Reference by name
    public let byName: String?
    
    /// Initialize an IR backreference
    ///
    /// - Parameters:
    ///   - byIndex: Optional numeric index
    ///   - byName: Optional name
    public init(byIndex: Int? = nil, byName: String? = nil) {
        self.byIndex = byIndex
        self.byName = byName
    }
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        var d: [String: Any] = ["ir": "Backref"]
        if let byIndex = byIndex {
            d["byIndex"] = byIndex
        }
        if let byName = byName {
            d["byName"] = byName
        }
        return d
    }
}

/// Represents a lookaround operation in the IR.
public struct IRLook {
    /// Direction: "Ahead" or "Behind"
    public let dir: String
    
    /// Whether the assertion is negative
    public let neg: Bool
    
    /// The body of the assertion
    public let body: IROp
    
    /// Initialize an IR lookaround
    ///
    /// - Parameters:
    ///   - dir: The direction ("Ahead" or "Behind")
    ///   - neg: Whether it's negative
    ///   - body: The assertion body
    public init(dir: String, neg: Bool, body: IROp) {
        self.dir = dir
        self.neg = neg
        self.body = body
    }
    
    /// Serialize to dictionary
    public func toDict() -> [String: Any] {
        return [
            "ir": "Look",
            "dir": dir,
            "neg": neg,
            "body": body.toDict()
        ]
    }
}

// MARK: - Codable Conformance

extension IRQuantMax: Codable {
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let intVal = try? container.decode(Int.self) {
            self = .count(intVal)
        } else if let strVal = try? container.decode(String.self), strVal == "Inf" {
            self = .inf
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Invalid IRQuantMax value")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .count(let n): try container.encode(n)
        case .inf: try container.encode("Inf")
        }
    }
}

extension IROp: Codable {
    enum CodingKeys: String, CodingKey {
        case ir
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .ir)

        switch type {
        case "Alt":
            self = .alt(try IRAlt(from: decoder))
        case "Seq":
            self = .seq(try IRSeq(from: decoder))
        case "Lit":
            self = .lit(try IRLit(from: decoder))
        case "Dot":
            self = .dot(try IRDot(from: decoder))
        case "Anchor":
            self = .anchor(try IRAnchor(from: decoder))
        case "CharClass":
            self = .charClass(try IRCharClass(from: decoder))
        case "Quant":
            self = .quant(try IRQuant(from: decoder))
        case "Group":
            self = .group(try IRGroup(from: decoder))
        case "Backref":
            self = .backref(try IRBackref(from: decoder))
        case "Look":
            self = .look(try IRLook(from: decoder))
        default:
            throw DecodingError.dataCorruptedError(forKey: .ir, in: container, debugDescription: "Unknown IR op type: \(type)")
        }
    }

    public func encode(to encoder: Encoder) throws {
        // Encoding implementation omitted
    }
}

extension IRAlt: Codable {
    enum CodingKeys: String, CodingKey {
        case branches
    }
}

extension IRSeq: Codable {
    enum CodingKeys: String, CodingKey {
        case parts
    }
}

extension IRLit: Codable {
    enum CodingKeys: String, CodingKey {
        case value
    }
}

extension IRDot: Codable {
    public init(from decoder: Decoder) throws {
        self.init()
    }

    public func encode(to encoder: Encoder) throws {
        // Omitted
    }
}

extension IRAnchor: Codable {
    enum CodingKeys: String, CodingKey {
        case at
    }
}

// Helper for decoding IRClassItems
struct AnyIRClassItem: Codable {
    let item: IRClassItem

    enum CodingKeys: String, CodingKey {
        case ir
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .ir)

        switch type {
        case "Range":
            self.item = try IRClassRange(from: decoder)
        case "Char":
            self.item = try IRClassLiteral(from: decoder)
        case "Esc":
            self.item = try IRClassEscape(from: decoder)
        default:
            throw DecodingError.dataCorruptedError(forKey: .ir, in: container, debugDescription: "Unknown IR class item type: \(type)")
        }
    }

    func encode(to encoder: Encoder) throws {
        // Omitted
    }
}

extension IRClassRange: Codable {
    enum CodingKeys: String, CodingKey {
        case fromCh = "from"
        case toCh = "to"
    }
}

extension IRClassLiteral: Codable {
    enum CodingKeys: String, CodingKey {
        case ch = "char"
    }
}

extension IRClassEscape: Codable {
    enum CodingKeys: String, CodingKey {
        case type
        case property
    }
}

extension IRCharClass: Codable {
    enum CodingKeys: String, CodingKey {
        case negated
        case items
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.negated = try container.decode(Bool.self, forKey: .negated)
        let anyItems = try container.decode([AnyIRClassItem].self, forKey: .items)
        self.items = anyItems.map { -bash.item }
    }

    public func encode(to encoder: Encoder) throws {
        // Omitted
    }
}

extension IRQuant: Codable {
    enum CodingKeys: String, CodingKey {
        case child
        case min
        case max
        case mode
    }
}

extension IRGroup: Codable {
    enum CodingKeys: String, CodingKey {
        case capturing
        case body
        case name
        case atomic
    }
}

extension IRBackref: Codable {
    enum CodingKeys: String, CodingKey {
        case byIndex
        case byName
    }
}

extension IRLook: Codable {
    enum CodingKeys: String, CodingKey {
        case dir
        case neg
        case body
    }
}

// Equatable conformance for testing
extension IROp: Equatable {
    public static func == (lhs: IROp, rhs: IROp) -> Bool {
        switch (lhs, rhs) {
        case (.alt(let l), .alt(let r)): return l == r
        case (.seq(let l), .seq(let r)): return l == r
        case (.lit(let l), .lit(let r)): return l == r
        case (.dot, .dot): return true
        case (.anchor(let l), .anchor(let r)): return l == r
        case (.charClass(let l), .charClass(let r)): return l == r
        case (.quant(let l), .quant(let r)): return l == r
        case (.group(let l), .group(let r)): return l == r
        case (.backref(let l), .backref(let r)): return l == r
        case (.look(let l), .look(let r)): return l == r
        default: return false
        }
    }
}

extension IRAlt: Equatable {
    public static func == (lhs: IRAlt, rhs: IRAlt) -> Bool {
        return lhs.branches == rhs.branches
    }
}

extension IRSeq: Equatable {
    public static func == (lhs: IRSeq, rhs: IRSeq) -> Bool {
        return lhs.parts == rhs.parts
    }
}

extension IRLit: Equatable {
    public static func == (lhs: IRLit, rhs: IRLit) -> Bool {
        return lhs.value == rhs.value
    }
}

extension IRAnchor: Equatable {
    public static func == (lhs: IRAnchor, rhs: IRAnchor) -> Bool {
        return lhs.at == rhs.at
    }
}

extension IRCharClass: Equatable {
    public static func == (lhs: IRCharClass, rhs: IRCharClass) -> Bool {
        guard lhs.negated == rhs.negated else { return false }
        guard lhs.items.count == rhs.items.count else { return false }
        // Simple check, assuming order matters or items are comparable
        // Since IRClassItem is a protocol, we can't easily equate arrays of them without type erasure or casting
        // For now, let's rely on string representation or assume they are same type and order
        for (i, item) in lhs.items.enumerated() {
            if !areEqual(item, rhs.items[i]) { return false }
        }
        return true
    }
}

func areEqual(_ lhs: IRClassItem, _ rhs: IRClassItem) -> Bool {
    if let l = lhs as? IRClassRange, let r = rhs as? IRClassRange { return l == r }
    if let l = lhs as? IRClassLiteral, let r = rhs as? IRClassLiteral { return l == r }
    if let l = lhs as? IRClassEscape, let r = rhs as? IRClassEscape { return l == r }
    return false
}

extension IRClassRange: Equatable {}
extension IRClassLiteral: Equatable {}
extension IRClassEscape: Equatable {}

extension IRQuant: Equatable {
    public static func == (lhs: IRQuant, rhs: IRQuant) -> Bool {
        return lhs.child == rhs.child && lhs.min == rhs.min && lhs.max == rhs.max && lhs.mode == rhs.mode
    }
}

extension IRQuantMax: Equatable {}

extension IRGroup: Equatable {
    public static func == (lhs: IRGroup, rhs: IRGroup) -> Bool {
        return lhs.capturing == rhs.capturing && lhs.body == rhs.body && lhs.name == rhs.name && lhs.atomic == rhs.atomic
    }
}

extension IRBackref: Equatable {
    public static func == (lhs: IRBackref, rhs: IRBackref) -> Bool {
        return lhs.byIndex == rhs.byIndex && lhs.byName == rhs.byName
    }
}

extension IRLook: Equatable {
    public static func == (lhs: IRLook, rhs: IRLook) -> Bool {
        return lhs.dir == rhs.dir && lhs.neg == rhs.neg && lhs.body == rhs.body
    }
}
